docs/proposals/25-private-cloud-profiles.md

```diff
@@ -0,0 +1,299 @@
+            - name: europe-central-1a
+            - name: europe-central-1b
+            - name: europe-central-1c
+        - name: europe-special
+          zones:
+            - name: europe-custom-1a
+            - name: europe-custom-1b
+            - name: europe-custom-1c
+```
+
+The rendering is done by a new custom controller registered to the `gardener-controller-manager`. It a merge conflict is to appear during the rendering process, it should be noted somewhere in the status and the parent cloud profile should be preferred.
+
+### Custom rbac verb
+
+To prevent users from entering arbitrary values in the `kubernetes` and `machineImages` fields, two custom rbac verbs may be introduced. It cloud then be checked if the user that is creating or updating the `kubernetes` or `machineImages` field is authorized to do so.
+
+```yaml
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: privatecloudprofile-modify-special-fields
+rules:
+  - apiGroups: [""]
+    resources: ["privatecloudprofile"]
+    verbs: ["modify-spec-kubernetes", "modify-spec-machineImages"]
+```
+
+### Automated removal of outdated Kubernetes versions and machine image versions
+
+Expired Kubernetes versions in cloud profiles are currently being removed manually. This is not an issue since there is a very limited amount of cloud profiles. However, it any project administrator is allowed to create private cloud profile which can then have its Kubernetes versions expiration dates extended, it cloud become a very cumbersome process to manually remove all the expired Kubernetes versions. It is therefore proposed to implement a custom controller managed my `gardener-controller-manager`, that reconciles private cloud profiles (and possibly even regular cloud profiles) and removes any Kubernetes versions that are past their expiration date. The same could be done for machine images versions.
+
+However, a problem arises as the kubernetes/machine image versions are referenced in the `providerConfig` and would therefore need to be removed manually anyways.
+
+### Cross project sharing
+
+A use case could be defined where a private cloud profile might want to shared across multiple projects and not just be used within the project it was created id. However this use case seems to be very slim so this functionality will not be implemented as of now. Although, when taking a broader view with Private Seeds and Cloud in Country, this feature is going to be necessary at some point. Still, is is not planned to be implemented as of this GEP.
+
+### Multi-inheritance
+
+Theoretically, a private cloud profile could inherit from a cloud profile that already inherits from a cloud profile. However, this should probably not be allowed since it presents some major challenges.
+
+1. The rendering process now has to merge multiple instead of two objects
+2. The parent field becomes more complex because it could point to a cloud profile or to a private cloud profile.
```

Comment:
This again sounds like multi-level inheritance. Why is that a problem? I don't see why this is bad (yet).

Multi-inheritance is bad, multi-level inheritance doesn't feel bad to me (yet).

Reply:
From my perspective, the biggest problem with multi-level inheritance is that the reconcile order of the (private) cloud profiles is now critical for correct rendering of the private cloud profiles. A change to a private cloud profile that both inherits and is inherited would need to communicate its change. Multi-level inheritance could however be implemented, but it is not planned as of this GEP as the development effort to implement it is, from our perspective, too high for the use case defined.

Reply:
Last seen revision? And since when is "recursive" a problem for developers?

Reply:
Clarified in private discussion.